File: backend/services/mpesa.py
import base64
from datetime import datetime
import httpx
from config import settings


class MpesaService:
    SANDBOX_URL = "https://sandbox.safaricom.co.ke"
    PRODUCTION_URL = "https://api.safaricom.co.ke"

    @property
    def base_url(self) -> str:
        return self.PRODUCTION_URL if settings.MPESA_ENV == "production" else self.SANDBOX_URL

    async def _get_access_token(self) -> str:
        credentials = base64.b64encode(
            f"{settings.MPESA_CONSUMER_KEY}:{settings.MPESA_CONSUMER_SECRET}".encode()
        ).decode()
        async with httpx.AsyncClient() as client:
            response = await client.get(
                f"{self.base_url}/oauth/v1/generate?grant_type=client_credentials",
                headers={"Authorization": f"Basic {credentials}"},
            )
            response.raise_for_status()
            return response.json()["access_token"]

    def _generate_password(self) -> tuple[str, str]:
        timestamp = datetime.now().strftime("%Y%m%d%H%M%S")
        raw = f"{settings.MPESA_SHORTCODE}{settings.MPESA_PASSKEY}{timestamp}"
        password = base64.b64encode(raw.encode()).decode()
        return password, timestamp

    async def stk_push(self, phone: str, amount: float, account_ref: str) -> dict:
        token = await self._get_access_token()
        password, timestamp = self._generate_password()
        payload = {
            "BusinessShortCode": settings.MPESA_SHORTCODE,
            "Password": password,
            "Timestamp": timestamp,
            "TransactionType": "CustomerPayBillOnline",
            "Amount": int(amount),
            "PartyA": phone,
            "PartyB": settings.MPESA_SHORTCODE,
            "PhoneNumber": phone,
            "CallBackURL": f"{settings.MPESA_CALLBACK_URL}/api/mpesa/callback",
            "AccountReference": account_ref,
            "TransactionDesc": f"Escrow deposit: {account_ref}",
        }
        async with httpx.AsyncClient() as client:
            response = await client.post(
                f"{self.base_url}/mpesa/stkpush/v1/processrequest",
                json=payload,
                headers={"Authorization": f"Bearer {token}"},
            )
            response.raise_for_status()
            return response.json()

    async def b2c_payout(self, phone: str, amount: float, remarks: str) -> dict:
        token = await self._get_access_token()
        payload = {
            "InitiatorName": settings.MPESA_B2C_INITIATOR,
            "SecurityCredential": settings.MPESA_B2C_PASSWORD,
            "CommandID": "BusinessPayment",
            "Amount": int(amount),
            "PartyA": settings.MPESA_SHORTCODE,
            "PartyB": phone,
            "Remarks": remarks,
            "QueueTimeOutURL": settings.MPESA_B2C_QUEUE_URL,
            "ResultURL": settings.MPESA_B2C_RESULT_URL,
        }
        async with httpx.AsyncClient() as client:
            response = await client.post(
                f"{self.base_url}/mpesa/b2c/v1/paymentrequest",
                json=payload,
                headers={"Authorization": f"Bearer {token}"},
            )
            response.raise_for_status()
            return response.json()


mpesa_service = MpesaService()

File: vault/src/mpesa.rs
use base64::{engine::general_purpose::STANDARD, Engine};
use reqwest::Client;
use secrecy::ExposeSecret;
use serde::{Deserialize, Serialize};
use time::OffsetDateTime;

use crate::config::Config;

pub struct MpesaClient {
    http: Client,
    config: Config,
}

#[derive(Deserialize)]
struct TokenResponse {
    access_token: String,
}

#[derive(Serialize)]
#[serde(rename_all = "PascalCase")]
struct StkPushRequest<'a> {
    business_short_code: &'a str,
    password: String,
    timestamp: String,
    transaction_type: &'static str,
    amount: u64,
    party_a: String,
    party_b: &'a str,
    phone_number: String,
    call_back_url: String,
    account_reference: &'a str,
    transaction_desc: &'a str,
}

#[derive(Deserialize, Debug)]
#[serde(rename_all = "PascalCase")]
pub struct StkPushResponse {
    pub checkout_request_id: String,
    pub response_code: String,
    pub response_description: String,
    pub customer_message: String,
}

impl MpesaClient {
    pub fn new(config: Config) -> Self {
        Self {
            http: Client::new(),
            config,
        }
    }

    async fn access_token(&self) -> anyhow::Result<String> {
        let creds = STANDARD.encode(format!(
            "{}:{}",
            self.config.mpesa_consumer_key.expose_secret(),
            self.config.mpesa_consumer_secret.expose_secret(),
        ));

        let res: TokenResponse = self
            .http
            .get(format!(
                "{}/oauth/v1/generate?grant_type=client_credentials",
                self.config.mpesa_base_url()
            ))
            .header("Authorization", format!("Basic {creds}"))
            .send()
            .await?
            .error_for_status()?
            .json()
            .await?;

        Ok(res.access_token)
    }

    pub async fn stk_push(
        &self,
        transaction_id: &str,
        phone: &str,
        amount: u64,
        description: &str,
    ) -> anyhow::Result<StkPushResponse> {
        let token = self.access_token().await?;
        let now = OffsetDateTime::now_utc();
        let timestamp = format!(
            "{:04}{:02}{:02}{:02}{:02}{:02}",
            now.year(),
            now.month() as u8,
            now.day(),
            now.hour(),
            now.minute(),
            now.second(),
        );

        let raw_password = format!(
            "{}{}{}",
            self.config.mpesa_shortcode,
            self.config.mpesa_passkey.expose_secret(),
            timestamp
        );
        let password = STANDARD.encode(raw_password);

        let normalized_phone = normalize_phone(phone);

        let body = StkPushRequest {
            business_short_code: &self.config.mpesa_shortcode,
            password,
            timestamp,
            transaction_type: "CustomerPayBillOnline",
            amount,
            party_a: normalized_phone.clone(),
            party_b: &self.config.mpesa_shortcode,
            phone_number: normalized_phone,
            call_back_url: format!("{}/payments/callback", self.config.vault_public_url),
            account_reference: transaction_id,
            transaction_desc: description,
        };

        let res: StkPushResponse = self
            .http
            .post(format!(
                "{}/mpesa/stkpush/v1/processrequest",
                self.config.mpesa_base_url()
            ))
            .bearer_auth(token)
            .json(&body)
            .send()
            .await?
            .error_for_status()?
            .json()
            .await?;

        Ok(res)
    }
}

fn normalize_phone(phone: &str) -> String {
    let digits: String = phone.chars().filter(|c| c.is_ascii_digit()).collect();
    if digits.starts_with("0") {
        format!("254{}", &digits[1..])
    } else if digits.starts_with("254") {
        digits
    } else {
        format!("254{digits}")
    }
}
